pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	watcherapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, versions []string, socketPath string) error
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path     string
+	handlers map[string]RegisterCallbackFn
+	stopCh   chan interface{}
+	fs       utilfs.Filesystem
+	watcher  *fsnotify.Watcher
```

Comment:
I dont think it is needed because this api will be invoked at kubelet start. At the time of a sub-module initialization, this api will be invoked and since kubelet module initializes in serial order, i dont think there could be a race issue.

Reply:
In that case we should make this module interface clear by:
- removing the `AddHandler` function
- Providing the handlers in the NewWatcher function

e.g:
```golang
type PluginHandler struct {
     handlerType string
     handlerCbkFn RegisterCallbackFn
}

func NewWatcher(sockDir string, handlers ...PluginHandler) Wacher
```

Reply:
IMO, keeping `NewWatcher` and `AddHandler` separate calls seems a little cleaner. `AddHandler` is expected to be invoked just after a sub-module init, while `NewWatcher` should be called at basic kubelet service initialization stage.

Reply:
In that case add a comment explaining the behavior.
But to be honest, "*We don't need a mutex here because you are not supposed to be calling this function after start*" seems exactly like a comment "*Should never happen*" that's just waiting to happen.

Reply:
+1 for explicit comment.